---
layout: post
title:  "scRNA-seq数据分析"
category: 'bioinformation'
tags:  
author: y570pc
image: 
---


## 比较

#### 注释信息下载

http://asia.ensembl.org/info/data/ftp/index.html

|项目|scRNA-seq|2|3|4|5|6|7|
|---|---|---|---|---|---|---|---|
|测序方法|SMART-seq<br>CEL-seq<br>SCRB-seq<br>DROP-seq|皮蛋瘦肉粥|蜂蜜小蛋糕|灌汤包|南瓜饼|肉末蛋羹|豆浆油条|
|测序平台|Fluidigm C1<br>Wafergen ICELL8<br>10X Genomics Chromium|笋干炒肉|箩卜炒肉|剁椒鱼头|葱油蛏子|风味蹄筋|珍珠丸子|
|晚餐|牛肉砂锅|虾皮炒海带|牛肉炒西芹|芝麻豆腐|香菇炒肉|土豆丝饼|叉烧肉|

Q: why PCA for RNA-Seq but tSNE for scRNA-seq?[01]

A: This has more to do with the goal of the techniques. PCA in bulk RNAseq is intended for QC to see if there are outliers. tSNE is used in scRNA-seq is used to find cell-types or other groups. You could use tSNE in place of PCA in bulk RNAseq, but since there are parameters to tweak and it's more computationally expensive there's no great benefit to do so.

Q: Why Does Rna-Seq Read Count Fit Poisson Distribution?

A: Just picture a process whereby you take the genome and choose a location at random to produce a read. This is a Poisson process. The same is true for RNAseq, only instead of a genome, you're choosing a read from the transcriptome. This is made a bit more complex by the fact that some transcripts are present at higher abundance than others, but at it's core, it's still a poisson process. In fact, you can think of a negative binomial distribution as a mixture of poisson distributions with different means, which is a reasonable model for sampling from RNA transcripts with different abundances.

## 参考资料

[01]. [Identifying cell populations with scRNASeq, Tallulah S.Andrews](https://www.sciencedirect.com/science/article/pii/S0098299717300493#bbib66)